File: apps/demo-be-golang-gin/internal/domain/attachment.go
package domain

import "time"

// MaxAttachmentSize is the maximum allowed attachment size in bytes (10 MB).
const MaxAttachmentSize = 10 * 1024 * 1024

// AllowedMIMETypes lists the supported file MIME types for attachments.
var AllowedMIMETypes = map[string]bool{
	"image/jpeg":      true,
	"image/png":       true,
	"application/pdf": true,
}

// Attachment represents a file attached to an expense entry.
type Attachment struct {
	ID          string    `gorm:"primaryKey" json:"id"`
	ExpenseID   string    `gorm:"not null;index" json:"expense_id"`
	Filename    string    `gorm:"not null" json:"filename"`
	ContentType string    `gorm:"not null" json:"content_type"`
	Size        int64     `gorm:"not null" json:"size"`
	URL         string    `json:"url"`
	CreatedAt   time.Time `json:"created_at"`
}

// ValidateMIMEType checks if the MIME type is allowed.
func ValidateMIMEType(contentType string) error {
	if !AllowedMIMETypes[contentType] {
		return NewUnsupportedMediaTypeError("unsupported file type: " + contentType)
	}
	return nil
}

// ValidateFileSize checks if the file size is within limits.
func ValidateFileSize(size int64) error {
	if size > MaxAttachmentSize {
		return NewFileTooLargeError("file size exceeds the 10 MB limit")
	}
	return nil
}

File: apps/demo-be-fsharp-giraffe/src/DemoBeFsgi/Domain/Attachment.fs
module DemoBeFsgi.Domain.Attachment

open System
open DemoBeFsgi.Domain.Types

type Attachment =
    { Id: Guid
      ExpenseId: Guid
      Filename: string
      ContentType: string
      FileSize: int64
      Data: byte[]
      Url: string
      CreatedAt: DateTime }

let supportedContentTypes = set [ "image/jpeg"; "image/png"; "application/pdf" ]

let maxFileSize = 10L * 1024L * 1024L // 10MB

let validateContentType (contentType: string) =
    if supportedContentTypes.Contains(contentType) then
        Ok contentType
    else
        Error(UnsupportedMediaType $"Unsupported content type: %s{contentType}")

let validateFileSize (size: int64) =
    if size > maxFileSize then
        Error(FileTooLarge maxFileSize)
    else
        Ok size
